pallets/token_migration/src/lib.rs

```diff
@@ -198,17 +314,299 @@ impl<T: Trait> Module<T> {
+
+        // A storage optimization would be to combine bonuses with same offsets but
+        // this is a rare occurrence in practice and thus not worth paying the O(n) cost.
+
+        Self::update_bonus(&to, bonus);
+        Self::deposit_event(RawEvent::VestingBonusAdded(from, to, amount, start));
+        Ok(())
+    }
+
+    /// Unlock any vesting bonuses that can be unlocked for an account
+    fn unlock_vesting_bonus(who: T::AccountId) -> dispatch::DispatchResult {
+        let mut bonus = Self::bonus(&who).ok_or(Error::<T>::NoBonus)?;
+        let bonuses = &mut bonus.vesting_bonuses;
+        ensure!(!bonuses.is_empty(), Error::<T>::NoVestingBonus);
+
+        let now = <frame_system::Module<T>>::block_number();
+        let now_plus_1 = now + T::BlockNumber::from(1);
+
+        let vesting_duration = T::BlockNumber::from(T::VestingDuration::get());
+        let vesting_milestones = T::BlockNumber::from(T::VestingMilestones::get() as u32);
+        let milestone_as_bal = BalanceOf::<T>::from(T::VestingMilestones::get() as u32);
+
+        // XXX: The following division needs to be done only once in practice as vesting duration will be fixed.
+        let milestone_duration = vesting_duration / vesting_milestones;
+
+        let mut bonus_to_unlock = BalanceOf::<T>::zero();
+
+        let mut i = 0;
+        let mut completely_vested_bonus_indices = Vec::<usize>::new();
+        while i < bonuses.len() {
+            let start = bonuses[i].2;
+            let total_bonus = bonuses[i].0;
+            let locked_bonus = bonuses[i].1;
+            // start + vesting_duration - 1 <= now
+            if (start + vesting_duration) <= now_plus_1 {
+                // All remaining locked bonus has vested. Unlock
+                bonus_to_unlock += locked_bonus;
+                completely_vested_bonus_indices.push(i);
+            } else {
+                // Calculate number of milestones already passed
+                // milestones_passed = now - start + 1
+                let milestones_passed = (now_plus_1.saturating_sub(start)) / milestone_duration;
+                if milestones_passed == vesting_milestones {
+                    // Unlock all bonus in last milestone
+                    bonus_to_unlock += locked_bonus;
+                    completely_vested_bonus_indices.push(i);
+                } else {
+                    let bonus_per_milestone = total_bonus / milestone_as_bal;
+                    let bonus_to_be_unlocked_till_now =
+                        T::BlockNumberToBalance::convert(milestones_passed) * bonus_per_milestone;
+                    let expected_locked = total_bonus.saturating_sub(bonus_to_be_unlocked_till_now);
+                    if expected_locked < locked_bonus {
+                        bonus_to_unlock += locked_bonus - expected_locked;
+                        bonuses[i].1 = expected_locked
+                    }
+                }
+            }
+            i += 1;
+        }
```

Comment:
This block gets a bit hairy. Want me to attempt to refactor for readability?

Reply:
Does hairy mean long? If not, can you elaborate?

Reply:
The nested branches and imperative style makes intent difficult to discern. And yes, it is a bit long.

Reply:
Removed an extra branch. Added note about why it was there.